sql/core/src/main/scala/org/apache/spark/sql/expressions/UserDefinedFunction.scala

```diff
@@ -85,8 +94,9 @@ case class UserDefinedFunction protected[sql] (
    * @since 2.3.0
    */
   def withName(name: String): this.type = {
-    this._nameOption = Option(name)
-    this
+    val udf = copyAll()
+    udf._nameOption = Option(name)
```

Comment:
yea, I know. I just meant we added an interface `newInstance(name, nullable, determinism)` there.